compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -0,0 +1,254 @@
+    forM_ existingTemplates $ \template ->
+        withContext
+            (ContextTemplate (present module_) (present template) TPWhole)
+            (checkTemplate module_ template)
+
+    -- TODO: Handle deleted datatypes (not belonging to templates or choices)
+    let (_dtDel, dtExisting, _dtNew) = extractDelExistNew $ NM.toHashMap . moduleDataTypes <$> module_
+
+    -- For a datatype, derive its context
+    let deriveChoiceInfo :: LF.Module -> HMS.HashMap LF.TypeConName (LF.Template, LF.TemplateChoice)
+        deriveChoiceInfo module_ = HMS.fromList $ do
+            template <- NM.toList (moduleTemplates module_)
+            choice <- NM.toList (tplChoices template)
+            TCon dtName <- [snd (chcArgBinder choice)] -- Choice inputs should always be a list
+            pure (qualObject dtName, (template, choice))
+        allChoiceReturnTCons :: LF.Module -> HMS.HashMap LF.TypeConName (LF.Template, LF.TemplateChoice)
+        allChoiceReturnTCons module_ = HMS.fromList $ do
+            template <- NM.toList (moduleTemplates module_)
+            choice <- NM.toList (tplChoices template)
+            dtName <- flip cata (chcReturnType choice) $ \case
+                TConF dtName -> [dtName]
+                rest -> fold rest
+            pure (qualObject dtName, (template, choice))
+        dataTypeOrigin
+            :: DefDataType -> Module
+            -> (UpgradedRecordOrigin, Context)
+        dataTypeOrigin dt module_
+            | Just template <- NM.name dt `NM.lookup` moduleTemplates module_ =
+                ( TemplateBody (NM.name dt)
+                , ContextTemplate module_ template TPWhole
+                )
+            | Just (template, choice) <- NM.name dt `HMS.lookup` deriveChoiceInfo module_ =
+                ( TemplateChoiceInput (NM.name template) (NM.name choice)
+                , ContextTemplate module_ template (TPChoice choice)
+                )
+            | Just (template, choice) <- NM.name dt `HMS.lookup` allChoiceReturnTCons module_ =
+                ( TemplateChoiceOutput (NM.name template) (NM.name choice) (NM.name dt)
+                , ContextTemplate module_ template (TPChoice choice)
+                )
+            | otherwise = (TopLevel, ContextDefDataType module_ dt)
+
+    forM_ dtExisting $ \dt ->
+        -- Get origin/context for each datatype in both past and present
+        let origin = dataTypeOrigin <$> dt <*> module_
+        in
+        -- If origins don't match, record has changed origin
+        if foldU (/=) (fst <$> origin) then
+            withContext (ContextDefDataType (present module_) (present dt)) $
+                throwWithContext (EUpgradeRecordChangedOrigin (dataTypeCon (present dt)) (fst (past origin)) (fst (present origin)))
+        else
+            case present origin of
+              (TopLevel, _) -> pure () -- Ignore top level datatypes
```

Comment:
We're saying top level data types can't be deleted or moved, but can be modified in non compatible ways?

Reply:
Good point - changing that to a TODO.